pkg/kubectl/cmd/apply.go

```diff
@@ -360,174 +330,25 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out, errOut io.Writer, opti
-			{"extensions", "v1beta1", "ReplicaSet", true},
-			{"apps", "v1beta1", "StatefulSet", true},
-			{"apps", "v1beta1", "Deployment", true},
-		}
-	}
-
-	for _, resource := range *pruneResources {
-		addedMapping, err := mapper.RESTMapping(schema.GroupKind{Group: resource.group, Kind: resource.kind}, resource.version)
-		if err != nil {
-			return nil, nil, fmt.Errorf("invalid resource %v: %v", resource, err)
-		}
-		if resource.namespaced {
-			namespaced = append(namespaced, addedMapping)
-		} else {
-			nonNamespaced = append(nonNamespaced, addedMapping)
-		}
-	}
-
-	return namespaced, nonNamespaced, nil
-}
-
-type pruner struct {
-	mapper        meta.RESTMapper
-	clientFunc    resource.ClientMapperFunc
-	clientsetFunc func() (internalclientset.Interface, error)
-
-	visitedUids   sets.String
-	labelSelector string
-	fieldSelector string
-
-	cascade     bool
-	dryRun      bool
-	gracePeriod int
-
-	out io.Writer
-}
-
-func (p *pruner) prune(f cmdutil.Factory, namespace string, mapping *meta.RESTMapping, shortOutput, includeUninitialized bool) error {
-	c, err := p.clientFunc(mapping)
-	if err != nil {
-		return err
-	}
-
-	objList, err := resource.NewHelper(c, mapping).List(
-		namespace,
-		mapping.GroupVersionKind.Version,
-		false,
-		&metav1.ListOptions{
-			LabelSelector:        p.labelSelector,
-			FieldSelector:        p.fieldSelector,
+	// If prune, then delete all unvisited objects under "pruneGroupVersionKinds" mappings.
+	if options.Prune {
+		pruner := prune.Pruner{
+			Mapper:               mapper,
+			GroupVersionKinds:    pruneGroupVersionKinds,
+			ClientFunc:           f.UnstructuredClientForMapping,
+			ClientsetFunc:        f.ClientSet,
+			Selector:             options.Selector,
 			IncludeUninitialized: includeUninitialized,
-		},
-	)
-	if err != nil {
-		return err
-	}
-	objs, err := meta.ExtractList(objList)
-	if err != nil {
-		return err
-	}
-
-	for _, obj := range objs {
-		annots, err := mapping.MetadataAccessor.Annotations(obj)
-		if err != nil {
-			return err
-		}
-		if _, ok := annots[api.LastAppliedConfigAnnotation]; !ok {
-			// don't prune resources not created with apply
-			continue
-		}
-		uid, err := mapping.UID(obj)
-		if err != nil {
-			return err
-		}
-		if p.visitedUids.Has(string(uid)) {
-			continue
-		}
-
-		name, err := mapping.Name(obj)
-		if err != nil {
-			return err
-		}
-		if !p.dryRun {
-			if err := p.delete(namespace, name, mapping); err != nil {
-				return err
-			}
+			Cascade:              options.Cascade,
+			GracePeriod:          options.GracePeriod,
+			DryRun:               dryRun,
+			Out:                  out,
+			ShortOutput:          shortOutput,
 		}
-		f.PrintSuccess(p.mapper, shortOutput, p.out, mapping.Resource, name, p.dryRun, "pruned")
-	}
-	return nil
-}
-
-func (p *pruner) delete(namespace, name string, mapping *meta.RESTMapping) error {
-	c, err := p.clientFunc(mapping)
-	if err != nil {
-		return err
+		pruner.Prune(visitedNamespaces, visitedUids)
```

Comment:
Shouldn't you return the error returned by `pruner.Prune`?

Reply:
If yes, would it make sense (or is it possible) to improve the test so that you would have caught that?